clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,695 @@
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(input.value));
+        input.show();
+    }).finally(() => { input.dispose(); });
+}
+
+function pickUser<T>(title: string, options: { label: string, value: T }[]): Promise<T> {
+    const input = vscode.window.createQuickPick();
+    return new Promise<T>((resolve, reject) => {
+        input.title = title;
+        input.items = options.map(x => { return { label: x.label }; });
+        input.canSelectMany = false;
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(options.find(x => x.label === input.selectedItems[0].label)!.value));
+        input.show();
+    }).finally(() => { input.dispose(); });
+}
+
+// because VSCode does not expose the service???
+export function replaceVariables(obj: any, resolver: (configKey: string) => (string | undefined), worksapceUri: vscode.Uri): any {
+    if (typeof obj === 'object') {
+        for (const x in obj)
+            obj[x] = replaceVariables(obj[x], resolver, worksapceUri);
+    }
+    else if (typeof obj === 'string') {
+        while (true) {
+            const match = /\$\{config:([^}]+)\}|\$\{(workspaceFolder)\}/.exec(obj);
+            if (!match) break;
+
+            const replacement = match[1] ? '' + resolver(match[1]) : worksapceUri;
+            obj = obj.slice(0, match.index) + replacement + obj.slice(match.index + match[0].length);
+        }
+    }
+    return obj;
+}
+
+async function gatherAvailableConfigs() {
+    if (vscode.workspace.workspaceFolders === undefined) return [];
+    const availableConfigs = (await Promise.all(vscode.workspace.workspaceFolders.map(x => {
+        return new Promise<{ workspace: vscode.WorkspaceFolder, config: any } | null>((resolve) => {
+            vscode.workspace.openTextDocument(vscode.Uri.joinPath(x.uri, ".hlasmplugin", "proc_grps.json")).then((doc) => resolve({ workspace: x, config: JSON.parse(doc.getText()) }), _ => resolve(null))
+        })
+    }))).filter(x => !!x).map(x => x!);
+
+    const varResolver = (workspace: vscode.WorkspaceFolder) => {
+        const config = vscode.workspace.getConfiguration(undefined, workspace);
+        return (s: string) => config.get<string>(s);
+    }
+
+    return availableConfigs.map(x => { return { workspaceUri: x.workspace.uri, config: replaceVariables(x.config, varResolver(x.workspace), x.workspace.uri) } });
+}
+
+export function extractDsn(d: string | undefined, workspaceUri: vscode.Uri): { dsn: string, path: string } | null {
+    const guessDsnRegex = /(?:.*[\\/])?((?:[A-Za-z0-9@#$]{1,8})(?:\.[A-Za-z0-9@#$]{1,8})+)[\\/]*/;
+
+    if (!d || d.length === 0)
+        return null;
+
+    const dsn_match = guessDsnRegex.exec(d);
+    if (dsn_match) {
+        const dsn = dsn_match[1].toUpperCase();
+        if (d.startsWith("~"))
+            return { dsn: dsn, path: fixPath(path.join(homedir(), /~[\\/]/.test(d) ? d.slice(2) : d.slice(1))) };
```

Comment:
What is an example of the case when `/~[\\/]/.test(d)` is falsy and `d.slice(1)` is desired?

Reply:
That's for consistency with the server implementation - ~something/ should never really be used - we don't give it the unix/linux style interpretation.